src/main/scala/au/com/agiledigital/toolform/model/Project.scala

```diff
@@ -0,0 +1,69 @@
+package au.com.agiledigital.toolform.model
+
+import pureconfig.{ConfigConvert, ConfigFieldMapping, ProductHint}
+import pureconfig.ConfigConvert.viaStringOpt
+
+/**
+  * Configuration for a project.
+  *
+  * @param id         the id of the project.
+  * @param name       the name of the project.
+  * @param components the components that will be built and combined to form the projects.
+  * @param resources  the resources that must be supplied to the project to run.
+  * @param topology   the network topology of the environment explains how elements will be related/connected.
+  * @param volumes    the volumes that will be mounted into components.
+  */
+case class Project(id: String,
+                   name: String,
+                   components: Map[String, Component],
+                   resources: Map[String, Resource],
+                   topology: Topology,
+                   volumes: Option[Seq[Volume]],
+                   componentGroups: Option[Seq[ComponentGroup]])
+
+/**
+  * the network topology of the environment explains how elements will be related/connected.
+  * @param links the links between components, other components and resources.
+  * @param edges the edges that will make the components and resources available outside the project.
+  */
+case class Topology(links: Seq[Link], edges: Map[String, Edge])
+
+object Topology {
+  implicit val fieldMapping: ProductHint[Topology] = ProductHint[Topology](new ConfigFieldMapping {
+    def apply(fieldName: String): String = fieldName
+  })
+}
+
+/**
+  * A reference points to an element of the project that is defined elsewhere.
+  * See Link, Volume, Location.
+  * @param ref A dot separated path from the project root to the referenced object.  eg. "components.componentA"
+  */
+case class Reference(ref: String)
```

Comment:
Be nice if the type of the target element was reflected here.

Reply:
ADT

Reply:
Yeah, I wasn't sure if we should go straight to resolving the target during parsing vs. doing the resolution lazily (or on demand).

I think that might be a bit of a challenge to teach PureConfig how to do backrefs to the `Project` that it is actively unmarshalling so my tentative plan was to do so in a second pass.  I'll put up some code that shows this.